cozy/cost_model.py

```diff
@@ -93,16 +135,45 @@ def _compare(self, e1 : Exp, e2 : Exp, context : Context):
 
     def compare(self, e1 : Exp, e2 : Exp, context : Context, pool : Pool) -> Order:
         with task("compare costs", context=context):
-            if pool == RUNTIME_POOL:
-                return composite_order(
-                    lambda: order_objects(asymptotic_runtime(e1), asymptotic_runtime(e2)),
-                    lambda: self._compare(max_storage_size(e1, self.freebies), max_storage_size(e2, self.freebies), context),
-                    lambda: self._compare(rt(e1), rt(e2), context),
-                    lambda: order_objects(e1.size(), e2.size()))
+            if consider_maintenance_cost.value:
+                if pool == RUNTIME_POOL:
+                    return prioritized_order(
+                        lambda: order_objects(asymptotic_runtime(e1), asymptotic_runtime(e2)),
+                        lambda: unprioritized_order(
+                            [lambda: prioritized_order(
+                                lambda: self._compare(
+                                    max_storage_size(e1, self.freebies),
+                                    max_storage_size(e2, self.freebies), context),
+                                lambda: self._compare(rt(e1), rt(e2), context))] +
+                            [lambda op=op: self._compare(
+                                maintenance_cost(e1, self.solver, op, self.freebies),
+                                maintenance_cost(e2, self.solver, op, self.freebies),
+                                context) for op in self.ops]),
+                        lambda: order_objects(e1.size(), e2.size()))
+                else:
+                    return prioritized_order(
+                        lambda: unprioritized_order(
+                            [lambda: prioritized_order(
+                                lambda: self._compare(
+                                    max_storage_size(e1, self.freebies),
+                                    max_storage_size(e2, self.freebies), context),
+                                lambda: self._compare(rt(e1), rt(e2), context))] +
+                            [lambda op=op: self._compare(
+                                maintenance_cost(e1, self.solver, op, self.freebies),
+                                maintenance_cost(e2, self.solver, op, self.freebies),
+                                context) for op in self.ops]),
+                        lambda: order_objects(e1.size(), e2.size()))
```

Comment:
This whole rewritten code for the `pool != RUNTIME_POOL` case seems very off. Non-runtime expressions are those that _might_ become _part of_ a value that is stored on the data structure rather than recomputed.

I would suggest using exactly the simple "storage-size / ast-size" strategy a few lines down that the old code uses.

Problems with this rewritten approach:
 - The runtime `rt` should not factor into this cost at all. These expressions will never be executed.
 - These expressions will not contain `EStateVar` nodes, so their `maintenance_cost`s are always exactly zero.